### PR TITLE
Spike/38838 Add SSH support to azure docker images

### PR DESCRIPTION
**Pull request by GuyHarwood**
Adds SSH support to docker images from azure portal.

- enables direct access to containers from azure portal
- allows viewing of PM2 log output from within container via `pm2 logs`

### CODE REVIEW

File: pupil-api/Dockerfile [hunk 1]
```diff
@@ -10,5 +10,18 @@ COPY . .
...
 RUN yarn global add pm2
 
-EXPOSE 3003
+# ssh
+ENV SSH_PASSWD "root:Docker!"
+RUN apt-get update \
+        && apt-get install -y --no-install-recommends dialog \
+        && apt-get update \
+	&& apt-get install -y --no-install-recommends openssh-server \
+	&& echo "$SSH_PASSWD" | chpasswd
```
Comment (jon-shipley) [inline pupil-api/Dockerfile]:
```
RUN apt-get update \
    && apt-get install -y --no-install-recommends dialog openssh-server \
    && echo "$SSH_PASSWD" | chpasswd
```

File: pupil-api/docker-start.sh [hunk 2]
```diff
@@ -10,4 +10,8 @@ then
   yarn migrate-sql
 fi
 
+echo "Starting SSH..."
+service ssh start
```
Comment (jon-shipley) [inline pupil-api/docker-start.sh]: Again, I would start the `ssh` service early.  I have no idea why `docker-start` is running migrations from this container.  I think they can be removed.

File: pupil-api/sshd_config [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# This is ssh server systemwide configuration file.
+#
+# /etc/sshd_config
+
```
Comment (jon-shipley) [inline pupil-api/sshd_config]: Again, a note to say that this config should not be exposed publicly would welcome.

File: pupil-spa/sshd_config [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# This is ssh server systemwide configuration file.
+#
+# /etc/sshd_config
+
```
Comment (jon-shipley) [inline pupil-spa/sshd_config]: Same here as before.

File: admin/Dockerfile [hunk 5]
```diff
@@ -4,11 +4,20 @@ RUN mkdir -p /usr/src/app
...
 RUN yarn install --ignore-engines --frozen-lockfile
-# RUN ./node_modules/.bin/gulp build
 RUN yarn global add pm2
 
-EXPOSE 3001
+# ssh
+ENV SSH_PASSWD "root:Docker!"
+RUN apt-get update \
+        && apt-get install -y --no-install-recommends dialog \
+        && apt-get update \
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/Dockerfile]: this is taken verbatim from the MS documentation 😄 i think that should explain everything.

good catch, will refine 💪

File: admin/Dockerfile [hunk 6]
```diff
@@ -4,11 +4,20 @@ RUN mkdir -p /usr/src/app
...
+	&& apt-get install -y --no-install-recommends openssh-server \
+	&& echo "$SSH_PASSWD" | chpasswd
 
-RUN chmod +x wait-sqldb.sh
+COPY sshd_config /etc/ssh/
 
-CMD ./wait-sqldb.sh && ./docker-start.sh
+EXPOSE 3001 2222
+
+RUN chmod u+x ./docker-start.sh
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/Dockerfile]: as above, taken as-is from MS docs to see if i could get a POC working.

this file was already set as executable, prior to this PR, so should be fine.

will remove statement

File: admin/docker-start.sh [hunk 7]
```diff
@@ -12,4 +12,8 @@ then
   yarn seed-sql
 fi
 
-pm2-docker start pm2.json
\ No newline at end of file
+echo "Starting SSH..."
+service ssh start
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/docker-start.sh]: the migrations are only used in test scenarios (`docker-compose.web.yml`, `docker-compose.pupil-test.yml`, `docker-compose.admin-test.yml`) 
its far from desirable, and would rather they were removed, as im sure this was done when dan was exploring the full test suite being run in a full compose setup.

File: admin/sshd_config [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+ListenAddress 		0.0.0.0
+LoginGraceTime 		180
+X11Forwarding 		yes
+Ciphers aes128-cbc,3des-cbc,aes256-cbc,aes128-ctr,aes192-ctr,aes256-ctr
+MACs hmac-sha1,hmac-sha1-96
+StrictModes 		yes
+SyslogFacility 		DAEMON
+PasswordAuthentication 	yes
+PermitEmptyPasswords 	no
+PermitRootLogin 	yes
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/sshd_config]: correct.  only accessible from azure portal.  will add note and link to docs

File: pupil-api/Dockerfile (hunk 1 above)
Comment (GuyHarwood) [reply pupil-api/Dockerfile]: already done as requested above 👍

**Comment by GuyHarwood** [comment]
@jon-shipley all points addressed, ready for re-review

File: admin/sshd_config [hunk 9]
```diff
@@ -0,0 +1,20 @@
+# This is ssh server systemwide configuration file.
+#
+# /etc/sshd_config
+#
+# enables SSH from azure portal into running container, not publicly exposed.
+# see docs for full info...
+# https://docs.microsoft.com/en-us/azure/app-service/containers/configure-custom-container
```
Comment (jon-shipley) [inline admin/sshd_config]: 👍

File: pupil-api/sshd_config [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# This is ssh server systemwide configuration file.
+#
+# /etc/sshd_config
+#
+# enables SSH from azure portal into running container, not publicly exposed.
+# see docs for full info...
+# https://docs.microsoft.com/en-us/azure/app-service/containers/configure-custom-container
```
Comment (jon-shipley) [inline pupil-api/sshd_config]: 👍

File: pupil-spa/Dockerfile [hunk 11]
```diff
@@ -68,4 +68,18 @@ COPY --from=builder /ng-app/docker-start.sh /
...
 #RUN rm -rf /usr/share/nginx/html/assets
 
+# install ssh server
+RUN apk add openssh openrc \
+     && echo "root:Docker!" | chpasswd
+
+COPY sshd_config /etc/ssh/
+
+# auto start ssh on startup
+# RUN rc-update add sshd
```
Comment (jon-shipley) [inline pupil-spa/Dockerfile]: Is this comment useful?

File: pupil-spa/sshd_config [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# This is ssh server systemwide configuration file.
+#
+# /etc/sshd_config
+#
+# enables SSH from azure portal into running container, not publicly exposed.
+# see docs for full info...
+# https://docs.microsoft.com/en-us/azure/app-service/containers/configure-custom-container
```
Comment (jon-shipley) [inline pupil-spa/sshd_config]: 👍

File: pupil-spa/Dockerfile (hunk 11 above)
Comment (GuyHarwood) [reply pupil-spa/Dockerfile]: yes, im waiting on confirmation about how to integrate this in alpine, as currently the documented alpine best practice does not work, nor does microsofts approach.

Comment (jon-shipley) [reply pupil-spa/Dockerfile]: Aha

I was wondering whether I should point out that this is not the same as every other docker container and whether it wouldn't make more sense to use the same base image for ease of use and configuration.

Comment (GuyHarwood) [reply pupil-spa/Dockerfile]: yes, we can discuss.  I would prefer to move away from containers completely, as its less baggage to maintain.  Will see what MS come up with on the support call

**Pull request merged**